### PR TITLE
Removal of code duplication

### PR DESCRIPTION
**Pull request by AliceBalfanz**
This PR removes code duplication within dsio. No other changes have been made.

### CODE REVIEW

File: xcube/core/dsio.py [hunk 1]
```diff
@@ -420,7 +419,7 @@ def read(self, path: str, **kwargs) -> xr.Dataset:
             if 'region_name' in kwargs:
                 client_kwargs['region_name'] = kwargs.pop('region_name')
```
Comment (forman) [inline xcube/core/dsio.py]: What about the lines above? Why do they not apply to write()? Or why are they in read(), not in write()?

Please check and adapt code so processing of args & kwargs in similar in read() and write()

File: xcube/core/dsio.py [hunk 2]
```diff
@@ -556,21 +554,15 @@ def _get_path_or_store(path: str,
...
-        if url.port is not None:
-            client_kwargs['endpoint_url'] = f'{url.scheme}://{url.host}:{url.port}'
-        else:
-            client_kwargs['endpoint_url'] = f'{url.scheme}://{url.host}'
-        root = url.path
-        if root.startswith('/'):
-            root = root[1:]
+        endpoint_url, root = split_bucket_url(path)
+        if endpoint_url:
+            client_kwargs['endpoint_url'] = endpoint_url
```
Comment (forman) [inline xcube/core/dsio.py]: Please also check if the code at

* https://github.com/dcs4cop/xcube/blob/master/xcube/core/mldataset.py#L610-L620

can be replaced by a function call that also simplifies read()/write() here. I imagine a function that takes our args and kwargs and outputs an s3fs.S3Map instance for read and/or write access.

**Comment by AliceBalfanz** [comment]
Closing PR due to more changes planned based on chat with forman, PR will be reopened after changes have been made.

File: xcube/core/dsio.py [hunk 3]
```diff
@@ -538,37 +528,50 @@ def rimraf(path):
...
         s3 = s3fs.S3FileSystem(anon=anon_mode, client_kwargs=client_kwargs)
-        path_or_store = s3fs.S3Map(root=root, s3=s3, create=mode == "write")
-    return path_or_store, root, client_kwargs
+        if mode == "w":
+            if not path.startswith("s3://"):
+                root = f's3://{root}'
+            consolidated = True
+        if mode == "r":
```
Comment (forman) [inline xcube/core/dsio.py:563]:
```suggestion
        elif mode == "r":
```

File: xcube/core/mldataset.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 from xcube.constants import FORMAT_NAME_NETCDF4
 from xcube.constants import FORMAT_NAME_SCRIPT
 from xcube.constants import FORMAT_NAME_ZARR
-from xcube.core.dsio import guess_dataset_format
+from xcube.core.dsio import guess_dataset_format, get_client_kwargs
```
Comment (forman) [inline xcube/core/mldataset.py:15]:
```suggestion
from xcube.core.dsio import guess_dataset_format
from xcube.core.dsio import get_client_kwargs
```

File: xcube/core/dsio.py [hunk 5]
```diff
@@ -538,37 +528,50 @@ def rimraf(path):
...
-        root = url.path
-        if root.startswith('/'):
-            root = root[1:]
-        if mode == "write":
-            root = f's3://{root}'
+    if path.startswith("https://") or path.startswith("http://") or path.startswith("s3://"):
+        endpoint_url, root = split_bucket_url(path)
+        if endpoint_url:
+            client_kwargs['endpoint_url'] = endpoint_url
         s3 = s3fs.S3FileSystem(anon=anon_mode, client_kwargs=client_kwargs)
```
Comment (forman) [inline xcube/core/dsio.py:558]: We need to pass `key` and `secret` now, see master branch

File: xcube/core/dsio.py [hunk 6]
```diff
@@ -538,37 +528,50 @@ def rimraf(path):
...
+        client_kwargs = client_kwargs.pop('client_kwargs')
+    if 'endpoint_url' in client_kwargs:
+        client_kwargs['endpoint_url'] = client_kwargs.pop('endpoint_url')
+    if 'region_name' in client_kwargs:
+        client_kwargs['region_name'] = kwargs.pop('region_name')
+    if 'provider_access_key_id' in client_kwargs and 'provider_secret_access_key' in client_kwargs:
+        anon_mode = False
+        client_kwargs['aws_access_key_id'] = client_kwargs.pop('provider_access_key_id')
+        client_kwargs['aws_secret_access_key'] = client_kwargs.pop('provider_secret_access_key')
+    return client_kwargs, anon_mode
```
Comment (forman) [inline xcube/core/dsio.py:544]: `client_kwargs['aws_access_key_id']` and `client_kwargs['aws_secret_access_key']` no longer valid

File: test/core/test_mldataset.py [hunk 7]
```diff
@@ -8,7 +8,8 @@
 import s3fs
 import xarray as xr
 
-from xcube.core.mldataset import BaseMultiLevelDataset
+from xcube.core.dsio import write_cube
+from xcube.core.mldataset import BaseMultiLevelDataset, open_ml_dataset_from_object_storage
```
Comment (forman) [inline test/core/test_mldataset.py:12]:
```suggestion
from xcube.core.mldataset import open_ml_dataset_from_object_storage
from xcube.core.mldataset import BaseMultiLevelDataset
```

**Pull request merged**